File: scripts/kb_pipeline/summarise.py
"""Shared AI summary generation using Claude tool-use mode.

Generates multi-level summaries (executive, detailed, takeaways) for content
items. Used by all pipeline scripts: ingest and markdown ingestion.

The summary result dict can be stored directly into content_items.summary_data
(JSONB), with summary['executive'] denormalised to content_items.ai_summary.
"""

import os
from datetime import datetime, timezone

from .config import get_env

# Max content length for summary prompt (matches batch_generate_summaries.ts)
MAX_SUMMARY_CONTENT_LENGTH = 100_000

# Sonnet pricing (per token) for summary cost estimation
SONNET_INPUT_PRICE = 3.0 / 1_000_000
SONNET_OUTPUT_PRICE = 15.0 / 1_000_000


def generate_summary(title: str, content: str, content_type: str,
                     primary_domain: str = "unknown") -> dict | None:
    """Generate a multi-level AI summary using Claude tool-use mode.

    Returns a dict with 'executive', 'detailed', 'takeaways', plus token/cost
    metadata, or None if generation fails.
    """
    try:
        import anthropic
    except ImportError:
        print("  [Summary] WARNING: anthropic package not installed — skipping summary")
        return None

    api_key = os.environ.get("ANTHROPIC_API_KEY") or get_env().get("ANTHROPIC_API_KEY")
    if not api_key:
        print("  [Summary] WARNING: ANTHROPIC_API_KEY not set — skipping summary")
        return None

    model = os.environ.get("AI_SUMMARY_MODEL", "claude-sonnet-4-6")

    # Truncate content to avoid excessive token usage
    truncated_content = content[:MAX_SUMMARY_CONTENT_LENGTH] if len(content) > MAX_SUMMARY_CONTENT_LENGTH else content

    is_transcript = content_type in ("transcript", "podcast", "video")

    prompt = f"""You are summarising content for a knowledge base.
Content type: {content_type}
Title: {title}
Domain: {primary_domain}

Rules:
- 3 to 7 takeaways
- Use UK English
- Be specific and factual, not vague
{"- This is a transcript/podcast: capture the speaker's key arguments, viewpoints, and any debates or disagreements between speakers" if is_transcript else "- For articles/posts: focus on the thesis, evidence, and conclusions"}
- The executive summary should be self-contained and informative

Content to summarise:
{truncated_content}"""

    tool_definition = {
        "name": "return_summary",
        "description": "Return the generated summary",
        "input_schema": {
            "type": "object",
            "properties": {
                "executive": {
                    "type": "string",
                    "description": "Single sentence summary (max 150 chars)",
                },
                "detailed": {
                    "type": "string",
                    "description": "2-3 paragraph detailed summary",
                },
                "takeaways": {
                    "type": "array",
                    "items": {"type": "string"},
                    "description": "3-7 key takeaways",
                },
            },
            "required": ["executive", "detailed", "takeaways"],
        },
    }

    try:
        client = anthropic.Anthropic(api_key=api_key)
        response = client.messages.create(
            model=model,
            max_tokens=2000,
            tools=[tool_definition],
            tool_choice={"type": "tool", "name": "return_summary"},
            messages=[{"role": "user", "content": prompt}],
        )

        if response.stop_reason == "max_tokens":
            print("  [Summary] WARNING: Response truncated (max_tokens reached)")
            return None

        # Extract tool use result from response
        tool_result = None
        for block in response.content:
            if block.type == "tool_use" and block.name == "return_summary":
                tool_result = block.input
                break

        if not tool_result:
            print("  [Summary] WARNING: No tool_use block in response")
            return None

        if not tool_result.get("executive") or not tool_result.get("detailed") or not isinstance(tool_result.get("takeaways"), list):
            print("  [Summary] WARNING: Invalid summary structure returned")
            return None

        input_tokens = response.usage.input_tokens if response.usage else 0
        output_tokens = response.usage.output_tokens if response.usage else 0
        cost = (input_tokens * SONNET_INPUT_PRICE) + (output_tokens * SONNET_OUTPUT_PRICE)

        return {
            "executive": tool_result["executive"],
            "detailed": tool_result["detailed"],
            "takeaways": tool_result["takeaways"],
            "generated_at": datetime.now(timezone.utc).isoformat(),
            "model": model,
            "tokens_used": input_tokens + output_tokens,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
            "cost": cost,
        }

    except Exception as e:
        print(f"  [Summary] ERROR: {e}")
        return None

File: lib/ai/summarise.ts
/**
 * AI content summarisation.
 * Generates structured summaries (executive, detailed, takeaways) using Claude.
 */

import type { SupabaseClient } from '@supabase/supabase-js';
import type { Database } from '@/supabase/types/database.types';
import { getAnthropicClient, getAIModel } from '@/lib/anthropic';
import { extractToolResult } from '@/lib/ai-parse';
import { SummaryResponseSchema } from '@/lib/validation/ai-schemas';
import type { SummaryResponse } from '@/lib/validation/ai-schemas';
import { toJson } from '@/lib/validation/jsonb';
import type { SummaryData } from '@/types/content';
import { AIServiceError } from '@/lib/ai/errors';
import { assertSuccessfulStop } from '@/lib/ai/stop-reason';
import { logger } from '@/lib/logger';

// ──────────────────────────────────────────
// Types
// ──────────────────────────────────────────

/** @public */
export interface SummariseParams {
  supabase: SupabaseClient<Database>;
  itemId: string;
  force: boolean;
  /** Optional — batch scripts may not have a user context. */
  userId?: string;
}

export interface SummariseResult {
  summary_data: SummaryData;
}

/** @public */
export interface CallSummaryAIParams {
  content: string;
  title: string;
  contentType: string;
  domain: string;
}

/** @public */
export interface CallSummaryAIResult {
  summaryData: SummaryData;
  inputTokens: number;
  outputTokens: number;
}

// ──────────────────────────────────────────
// Constants
// ──────────────────────────────────────────

const MAX_CONTENT_LENGTH = 100_000;

// ──────────────────────────────────────────
// Tool schema
// ──────────────────────────────────────────

const SUMMARY_TOOL = {
  name: 'return_summary',
  description: 'Return the generated summary',
  // Grounding shape: forced_tool_strict (B-INV-35,
  // AI_TOUCHPOINT_GROUNDING['summarise.callSummaryAI']).
  strict: true,
  input_schema: {
    type: 'object' as const,
    properties: {
      executive: {
        type: 'string',
        description: 'Single sentence summary (max 150 chars)',
      },
      detailed: {
        type: 'string',
        description: '2-3 paragraph detailed summary',
      },
      takeaways: {
        type: 'array',
        items: { type: 'string' },
        description: '3-7 key takeaways',
      },
    },
    required: ['executive', 'detailed', 'takeaways'] as string[],
    additionalProperties: false,
  },
} as const;

// ──────────────────────────────────────────
// Pure AI call
// ──────────────────────────────────────────

/**
 * Pure AI function: call Claude to generate a structured summary.
 * No Supabase dependency — suitable for both the service layer and batch scripts.
 *
 * Content is truncated to MAX_CONTENT_LENGTH internally.
 *
 * @throws AIServiceError on truncated response or invalid structure
 */
export async function callSummaryAI(
  params: CallSummaryAIParams,
): Promise<CallSummaryAIResult> {
  const { title, contentType, domain } = params;

  // Truncate if very long
  const content =
    params.content.length > MAX_CONTENT_LENGTH
      ? params.content.slice(0, MAX_CONTENT_LENGTH)
      : params.content;

  // Build the prompt
  const isTranscript = ['transcript', 'podcast', 'video'].includes(contentType);

  const prompt = `You are summarising content for a knowledge base.
Content type: ${contentType}
Title: ${title}
Domain: ${domain}

Rules:
- 3 to 7 takeaways
- Use UK English
- Be specific and factual, not vague
${isTranscript ? "- This is a transcript/podcast: capture the speaker's key arguments, viewpoints, and any debates or disagreements between speakers" : '- For articles/posts: focus on the thesis, evidence, and conclusions'}
- The executive summary should be self-contained and informative

Content to summarise:
${content}`;

  // Call Claude API
  const client = getAnthropicClient();
  const model = getAIModel();

  const response = await client.messages.create({
    model,
    max_tokens: 2000,
    tools: [SUMMARY_TOOL],
    tool_choice: { type: 'tool' as const, name: 'return_summary' },
    messages: [{ role: 'user', content: prompt }],
  });

  // B-INV-36: surface refusal / max_tokens explicitly (log + throw), never
  // swallow them. extractToolResult below throws if no tool_use block is present.
  assertSuccessfulStop(response, 'summarise.callSummaryAI');

  const parsed = extractToolResult<SummaryResponse>(
    response,
    'return_summary',
    SummaryResponseSchema,
  );

  // Validate the parsed response
  if (
    !parsed.executive ||
    !parsed.detailed ||
    !Array.isArray(parsed.takeaways)
  ) {
    throw new AIServiceError(
      'Invalid summary structure returned by Claude',
      500,
    );
  }

  const inputTokens = response.usage?.input_tokens ?? 0;
  const outputTokens = response.usage?.output_tokens ?? 0;

  const summaryData: SummaryData = {
    executive: parsed.executive,
    detailed: parsed.detailed,
    takeaways: parsed.takeaways,
    generated_at: new Date().toISOString(),
    model,
    tokens_used: inputTokens + outputTokens,
  };

  return { summaryData, inputTokens, outputTokens };
}

// ──────────────────────────────────────────
// Service-layer function
// ──────────────────────────────────────────

/**
 * Generate an AI summary for a content item.
 * Fetches the item, calls Claude via callSummaryAI(), stores the result, and returns it.
 *
 * @throws AIServiceError for domain errors (404, 400, 409, 413, 500)
 */
export async function generateSummary(
  params: SummariseParams,
): Promise<SummariseResult> {
  const { supabase, itemId, force, userId } = params;

  // Fetch the content item
  const { data: item, error: fetchError } = await supabase
    .from('content_items')
    .select(
      'id, content, title, suggested_title, content_type, summary, primary_domain, summary_data',
    )
    .eq('id', itemId)
    .single();

  if (fetchError || !item) {
    throw new AIServiceError('Content item not found', 404);
  }

  // Check if summary already exists (unless force=true)
  if (item.summary_data && !force) {
    throw new AIServiceError(
      'Summary already exists. Pass force=true to regenerate.',
      409,
    );
  }

  if (!item.content?.trim()) {
    throw new AIServiceError('Content item has no content to summarise', 400);
  }

  const displayTitle = item.suggested_title || item.title || 'Untitled';
  const contentType = item.content_type || 'article';
  const domain = item.primary_domain || 'unknown';

  // Call the pure AI function
  const { summaryData } = await callSummaryAI({
    content: item.content,
    title: displayTitle,
    contentType,
    domain,
  });

  // Store in Supabase (also sync summary with the higher-quality executive)
  const updatePayload: Database['public']['Tables']['content_items']['Update'] =
    {
      summary_data: toJson(summaryData),
      summary: summaryData.executive,
    };
  if (userId) {
    updatePayload.updated_by = userId;
  }

  const { error: updateError } = await supabase
    .from('content_items')
    .update(updatePayload)
    .eq('id', itemId);

  if (updateError) {
    logger.error(
      { err: updateError, op: 'summarise.store', itemId },
      'Failed to store summary',
    );
    throw new AIServiceError('Summary generated but failed to store', 500);
  }

  return { summary_data: summaryData };
}
